File: betting-algorithm/src/calibration.py
"""
Calibration Engine for Self-Improving Betting Algorithm

Tracks prediction performance and automatically adjusts factor weights
to improve accuracy over time.

Author: AI Betting Algorithm v2.0
Date: 2026-02-11
"""

from typing import Dict, List, Optional, Any
from dataclasses import dataclass, field
from collections import defaultdict
import numpy as np


@dataclass
class PredictionRecord:
    """Record of a single prediction with outcome"""
    match_id: str
    predicted_outcome: str  # 'home', 'draw', 'away'
    actual_outcome: str
    predicted_probs: Dict[str, float]  # {'home': 0.5, 'draw': 0.3, 'away': 0.2}
    factors: Dict[str, Any]  # Factor contributions
    correct: bool
    timestamp: str

    def __post_init__(self):
        """Calculate if prediction was correct"""
        self.correct = (self.predicted_outcome == self.actual_outcome)


class CalibrationEngine:
    """
    Self-improving calibration system for betting algorithm.

    Tracks prediction accuracy per factor and adjusts weights to
    maximize overall prediction accuracy.

    Features:
    - Per-factor performance tracking
    - Gradual weight adjustment with learning rate
    - Protection against overfitting (max ±10% adjustment)
    - Minimum sample size requirement (20 predictions)
    """

    def __init__(self, learning_rate: float = 0.01, max_history: int = 1000):
        """
        Initialize calibration engine.

        Args:
            learning_rate: Weight adjustment speed (0.01 = 1% per calibration)
            max_history: Maximum predictions to store (older dropped)
        """
        self.learning_rate = learning_rate
        self.max_history = max_history
        self.prediction_history: List[PredictionRecord] = []
        self.factor_performance: Dict[str, Dict[str, float]] = defaultdict(lambda: {
            'correct': 0,
            'total': 0,
            'accuracy': 0.0,
            'contribution': 0.0
        })

    def record_prediction(
        self,
        match_id: str,
        factors: Dict,
        predicted_probs: Dict[str, float],
        predicted_outcome: str,
        actual_outcome: str,
        timestamp: Optional[str] = None
    ):
        """
        Record a prediction and its actual outcome.

        Args:
            match_id: Unique match identifier
            factors: Factor data from prediction (dict or FactorResult objects)
            predicted_probs: Predicted probabilities {'home': 0.5, ...}
            predicted_outcome: Predicted result ('home', 'draw', 'away')
            actual_outcome: Actual result ('home', 'draw', 'away')
            timestamp: ISO timestamp (defaults to current time)
        """
        from datetime import datetime
        from .models import FactorResult

        if timestamp is None:
            timestamp = datetime.now().isoformat()

        # Create prediction record
        record = PredictionRecord(
            match_id=match_id,
            predicted_outcome=predicted_outcome,
            actual_outcome=actual_outcome,
            predicted_probs=predicted_probs,
            factors=factors,
            correct=(predicted_outcome == actual_outcome),
            timestamp=timestamp
        )

        # Add to history (maintain max size)
        self.prediction_history.append(record)
        if len(self.prediction_history) > self.max_history:
            self.prediction_history.pop(0)

        # Update factor performance tracking
        for factor_name, factor_data in factors.items():
            # Handle FactorResult objects
            if isinstance(factor_data, FactorResult):
                if factor_data.triggered:
                    self.factor_performance[factor_name]['total'] += 1
                    if record.correct:
                        self.factor_performance[factor_name]['correct'] += 1
                    self.factor_performance[factor_name]['contribution'] = factor_data.weight
            # Handle legacy dict factors
            elif isinstance(factor_data, dict):
                self.factor_performance[factor_name]['total'] += 1
                if record.correct:
                    self.factor_performance[factor_name]['correct'] += 1

        # Recalculate accuracy for all factors
        for factor_name, stats in self.factor_performance.items():
            if stats['total'] > 0:
                stats['accuracy'] = stats['correct'] / stats['total']

    def calibrate_weights(self, current_weights: Dict[str, float]) -> Dict[str, float]:
        """
        Adjust weights based on factor performance.

        Strategy:
        - Factors with above-average accuracy get slight weight increase
        - Factors with below-average accuracy get slight weight decrease
        - Adjustments capped at ±10% to prevent overfitting
        - Requires minimum 20 predictions

        Args:
            current_weights: Current factor weights

        Returns:
            Adjusted weights (normalized to sum = 1.0)
        """
        if len(self.prediction_history) < 20:
            # Not enough data for calibration
            return current_weights

        # Calculate overall accuracy
        overall_accuracy = sum(1 for r in self.prediction_history if r.correct) / len(self.prediction_history)

        if overall_accuracy == 0:
            # All predictions wrong, no calibration possible
            return current_weights

        # Calculate adjustment for each factor
        adjusted_weights = current_weights.copy()

        for factor_name, weight in current_weights.items():
            if factor_name not in self.factor_performance:
                continue

            stats = self.factor_performance[factor_name]
            if stats['total'] < 5:
                # Not enough data for this factor
                continue

            factor_accuracy = stats['accuracy']

            # Calculate relative performance
            performance_ratio = factor_accuracy / overall_accuracy

            # Calculate adjustment (positive if factor outperforms, negative if underperforms)
            adjustment = self.learning_rate * (performance_ratio - 1.0)

            # Cap adjustment at ±10%
            adjustment = max(-0.10, min(0.10, adjustment))

            # Apply adjustment
            new_weight = weight * (1 + adjustment)

            # Ensure weight stays positive
            adjusted_weights[factor_name] = max(0.01, new_weight)

        # Normalize weights to sum = 1.0
        total_weight = sum(adjusted_weights.values())
        if total_weight > 0:
            adjusted_weights = {k: v / total_weight for k, v in adjusted_weights.items()}

        return adjusted_weights

    def get_performance_report(self) -> Dict[str, Any]:
        """
        Generate comprehensive performance report.

        Returns:
            Dict with overall stats, factor stats, recent trends
        """
        if not self.prediction_history:
            return {
                'overall_accuracy': 0.0,
                'total_predictions': 0,
                'factor_stats': {},
                'recent_trend': 'N/A'
            }

        # Overall stats
        total_predictions = len(self.prediction_history)
        correct_predictions = sum(1 for r in self.prediction_history if r.correct)
        overall_accuracy = correct_predictions / total_predictions

        # Recent trend (last 20 predictions vs overall)
        recent_predictions = self.prediction_history[-20:]
        recent_correct = sum(1 for r in recent_predictions if r.correct)
        recent_accuracy = recent_correct / len(recent_predictions)

        trend = "improving" if recent_accuracy > overall_accuracy else "declining" if recent_accuracy < overall_accuracy else "stable"

        # Factor stats
        factor_stats = {}
        for factor_name, stats in self.factor_performance.items():
            if stats['total'] > 0:
                factor_stats[factor_name] = {
                    'accuracy': stats['accuracy'],
                    'total_samples': stats['total'],
                    'contribution': stats['contribution'],
                    'relative_performance': stats['accuracy'] / overall_accuracy if overall_accuracy > 0 else 1.0
                }

        # Outcome distribution
        outcome_counts = defaultdict(int)
        for record in self.prediction_history:
            outcome_counts[record.actual_outcome] += 1

        return {
            'overall_accuracy': overall_accuracy,
            'total_predictions': total_predictions,
            'correct_predictions': correct_predictions,
            'recent_accuracy': recent_accuracy,
            'recent_trend': trend,
            'factor_stats': factor_stats,
            'outcome_distribution': dict(outcome_counts),
            'sample_size_sufficient': total_predictions >= 20
        }

    def get_underperforming_factors(self, threshold: float = 0.9) -> List[str]:
        """
        Identify factors performing below threshold.

        Args:
            threshold: Performance ratio threshold (0.9 = 90% of overall accuracy)

        Returns:
            List of factor names performing below threshold
        """
        if len(self.prediction_history) < 20:
            return []

        overall_accuracy = sum(1 for r in self.prediction_history if r.correct) / len(self.prediction_history)

        if overall_accuracy == 0:
            return []

        underperforming = []
        for factor_name, stats in self.factor_performance.items():
            if stats['total'] >= 5:  # Minimum sample size
                relative_performance = stats['accuracy'] / overall_accuracy
                if relative_performance < threshold:
                    underperforming.append(factor_name)

        return underperforming

    def reset(self):
        """Clear all calibration data"""
        self.prediction_history.clear()
        self.factor_performance.clear()

    def export_history(self) -> List[Dict]:
        """
        Export prediction history as list of dicts.

        Returns:
            List of prediction records as dictionaries
        """
        return [
            {
                'match_id': r.match_id,
                'predicted_outcome': r.predicted_outcome,
                'actual_outcome': r.actual_outcome,
                'correct': r.correct,
                'predicted_probs': r.predicted_probs,
                'timestamp': r.timestamp
            }
            for r in self.prediction_history
        ]

    def import_history(self, records: List[Dict]):
        """
        Import prediction history from list of dicts.

        Args:
            records: List of prediction record dictionaries
        """
        for record in records:
            self.record_prediction(
                match_id=record['match_id'],
                factors={},  # Factors not stored in export
                predicted_probs=record['predicted_probs'],
                predicted_outcome=record['predicted_outcome'],
                actual_outcome=record['actual_outcome'],
                timestamp=record['timestamp']
            )

File: betting-algorithm-go/internal/algorithm/calibration.go
package algorithm

import (
	"sync"

	"bet4me/betting-algorithm-go/internal/domain"
)

// PredictionRecord represents a single prediction with its outcome.
type PredictionRecord struct {
	MatchID          string
	PredictedOutcome string
	ActualOutcome    string
	PredictedProbs   map[string]float64
	Factors          map[string]interface{}
	Correct          bool
	Timestamp        string
}

// CalibrationEngine is a self-improving calibration system for the betting algorithm.
//
// Tracks prediction accuracy per factor and adjusts weights to maximize overall accuracy.
//
// Features:
// - Per-factor performance tracking
// - Gradual weight adjustment with learning rate
// - Protection against overfitting (max ±10% adjustment)
// - Minimum sample size requirement (20 predictions)
// - Thread-safe with RWMutex for concurrent access
type CalibrationEngine struct {
	mu                 sync.RWMutex
	learningRate       float64
	maxHistory         int
	predictionHistory  []PredictionRecord
	factorPerformance  map[string]*FactorStats
}

// FactorStats tracks performance statistics for a single factor.
type FactorStats struct {
	Correct      int
	Total        int
	Accuracy     float64
	Contribution float64
}

// NewCalibrationEngine creates a new calibration engine.
func NewCalibrationEngine(learningRate float64, maxHistory int) *CalibrationEngine {
	return &CalibrationEngine{
		learningRate:      learningRate,
		maxHistory:        maxHistory,
		predictionHistory: make([]PredictionRecord, 0, maxHistory),
		factorPerformance: make(map[string]*FactorStats),
	}
}

// RecordPrediction records a prediction and its actual outcome.
//
// This method is thread-safe.
func (ce *CalibrationEngine) RecordPrediction(
	matchID string,
	factors map[string]interface{},
	predictedProbs map[string]float64,
	predictedOutcome string,
	actualOutcome string,
	timestamp string,
) {
	ce.mu.Lock()
	defer ce.mu.Unlock()

	// Create prediction record
	record := PredictionRecord{
		MatchID:          matchID,
		PredictedOutcome: predictedOutcome,
		ActualOutcome:    actualOutcome,
		PredictedProbs:   copyProbs(predictedProbs),
		Factors:          factors,
		Correct:          predictedOutcome == actualOutcome,
		Timestamp:        timestamp,
	}

	// Add to history (maintain max size)
	ce.predictionHistory = append(ce.predictionHistory, record)
	if len(ce.predictionHistory) > ce.maxHistory {
		ce.predictionHistory = ce.predictionHistory[1:]
	}

	// Update factor performance tracking
	for factorName, factorData := range factors {
		// Get or create stats
		if _, exists := ce.factorPerformance[factorName]; !exists {
			ce.factorPerformance[factorName] = &FactorStats{}
		}
		stats := ce.factorPerformance[factorName]

		// Handle FactorResult objects
		if fr, ok := factorData.(domain.FactorResult); ok {
			if fr.Triggered {
				stats.Total++
				if record.Correct {
					stats.Correct++
				}
				stats.Contribution = fr.Weight
			}
		} else if _, ok := factorData.(map[string]interface{}); ok {
			// Handle legacy dict factors
			stats.Total++
			if record.Correct {
				stats.Correct++
			}
		}
	}

	// Recalculate accuracy for all factors
	for _, stats := range ce.factorPerformance {
		if stats.Total > 0 {
			stats.Accuracy = float64(stats.Correct) / float64(stats.Total)
		}
	}
}

// CalibrateWeights adjusts weights based on factor performance.
//
// Strategy:
// - Factors with above-average accuracy get slight weight increase
// - Factors with below-average accuracy get slight weight decrease
// - Adjustments capped at ±10% to prevent overfitting
// - Requires minimum 20 predictions
//
// Returns adjusted weights (normalized to sum = 1.0).
// This method is thread-safe.
func (ce *CalibrationEngine) CalibrateWeights(currentWeights map[string]float64) map[string]float64 {
	ce.mu.RLock()
	defer ce.mu.RUnlock()

	if len(ce.predictionHistory) < 20 {
		// Not enough data for calibration
		return copyWeights(currentWeights)
	}

	// Calculate overall accuracy
	correct := 0
	for _, record := range ce.predictionHistory {
		if record.Correct {
			correct++
		}
	}
	overallAccuracy := float64(correct) / float64(len(ce.predictionHistory))

	if overallAccuracy == 0 {
		// All predictions wrong, no calibration possible
		return copyWeights(currentWeights)
	}

	// Calculate adjustment for each factor
	adjustedWeights := copyWeights(currentWeights)

	for factorName, weight := range currentWeights {
		stats, exists := ce.factorPerformance[factorName]
		if !exists || stats.Total < 5 {
			// Not enough data for this factor
			continue
		}

		factorAccuracy := stats.Accuracy

		// Calculate relative performance
		performanceRatio := factorAccuracy / overallAccuracy

		// Calculate adjustment (positive if factor outperforms, negative if underperforms)
		adjustment := ce.learningRate * (performanceRatio - 1.0)

		// Cap adjustment at ±10%
		if adjustment > 0.10 {
			adjustment = 0.10
		} else if adjustment < -0.10 {
			adjustment = -0.10
		}

		// Apply adjustment
		newWeight := weight * (1 + adjustment)

		// Ensure weight stays positive
		if newWeight < 0.01 {
			newWeight = 0.01
		}

		adjustedWeights[factorName] = newWeight
	}

	// Normalize weights to sum = 1.0
	total := 0.0
	for _, w := range adjustedWeights {
		total += w
	}

	if total > 0 {
		for k := range adjustedWeights {
			adjustedWeights[k] /= total
		}
	}

	return adjustedWeights
}

// GetPerformanceReport generates comprehensive performance report.
//
// This method is thread-safe.
func (ce *CalibrationEngine) GetPerformanceReport() map[string]interface{} {
	ce.mu.RLock()
	defer ce.mu.RUnlock()

	if len(ce.predictionHistory) == 0 {
		return map[string]interface{}{
			"overall_accuracy":        0.0,
			"total_predictions":       0,
			"factor_stats":            map[string]interface{}{},
			"recent_trend":            "N/A",
			"sample_size_sufficient":  false,
		}
	}

	// Overall stats
	totalPredictions := len(ce.predictionHistory)
	correctPredictions := 0
	for _, record := range ce.predictionHistory {
		if record.Correct {
			correctPredictions++
		}
	}
	overallAccuracy := float64(correctPredictions) / float64(totalPredictions)

	// Recent trend (last 20 predictions vs overall)
	recentPredictions := ce.predictionHistory
	if len(recentPredictions) > 20 {
		recentPredictions = recentPredictions[len(recentPredictions)-20:]
	}

	recentCorrect := 0
	for _, record := range recentPredictions {
		if record.Correct {
			recentCorrect++
		}
	}
	recentAccuracy := float64(recentCorrect) / float64(len(recentPredictions))

	trend := "stable"
	if recentAccuracy > overallAccuracy {
		trend = "improving"
	} else if recentAccuracy < overallAccuracy {
		trend = "declining"
	}

	// Factor stats
	factorStats := make(map[string]interface{})
	for factorName, stats := range ce.factorPerformance {
		if stats.Total > 0 {
			relativePerformance := 1.0
			if overallAccuracy > 0 {
				relativePerformance = stats.Accuracy / overallAccuracy
			}

			factorStats[factorName] = map[string]interface{}{
				"accuracy":             stats.Accuracy,
				"total_samples":        stats.Total,
				"contribution":         stats.Contribution,
				"relative_performance": relativePerformance,
			}
		}
	}

	// Outcome distribution
	outcomeDistribution := make(map[string]int)
	for _, record := range ce.predictionHistory {
		outcomeDistribution[record.ActualOutcome]++
	}

	return map[string]interface{}{
		"overall_accuracy":        overallAccuracy,
		"total_predictions":       totalPredictions,
		"correct_predictions":     correctPredictions,
		"recent_accuracy":         recentAccuracy,
		"recent_trend":            trend,
		"factor_stats":            factorStats,
		"outcome_distribution":    outcomeDistribution,
		"sample_size_sufficient":  totalPredictions >= 20,
	}
}

// GetUnderperformingFactors identifies factors performing below threshold.
//
// Returns list of factor names performing below the threshold.
// This method is thread-safe.
func (ce *CalibrationEngine) GetUnderperformingFactors(threshold float64) []string {
	ce.mu.RLock()
	defer ce.mu.RUnlock()

	if len(ce.predictionHistory) < 20 {
		return []string{}
	}

	correct := 0
	for _, record := range ce.predictionHistory {
		if record.Correct {
			correct++
		}
	}
	overallAccuracy := float64(correct) / float64(len(ce.predictionHistory))

	if overallAccuracy == 0 {
		return []string{}
	}

	underperforming := []string{}
	for factorName, stats := range ce.factorPerformance {
		if stats.Total >= 5 { // Minimum sample size
			relativePerformance := stats.Accuracy / overallAccuracy
			if relativePerformance < threshold {
				underperforming = append(underperforming, factorName)
			}
		}
	}

	return underperforming
}

// Reset clears all calibration data.
//
// This method is thread-safe.
func (ce *CalibrationEngine) Reset() {
	ce.mu.Lock()
	defer ce.mu.Unlock()

	ce.predictionHistory = make([]PredictionRecord, 0, ce.maxHistory)
	ce.factorPerformance = make(map[string]*FactorStats)
}

// ExportHistory exports prediction history as a list of simplified records.
//
// This method is thread-safe.
func (ce *CalibrationEngine) ExportHistory() []map[string]interface{} {
	ce.mu.RLock()
	defer ce.mu.RUnlock()

	result := make([]map[string]interface{}, len(ce.predictionHistory))
	for i, record := range ce.predictionHistory {
		result[i] = map[string]interface{}{
			"match_id":          record.MatchID,
			"predicted_outcome": record.PredictedOutcome,
			"actual_outcome":    record.ActualOutcome,
			"correct":           record.Correct,
			"predicted_probs":   copyProbs(record.PredictedProbs),
			"timestamp":         record.Timestamp,
		}
	}

	return result
}

// Helper functions

func copyWeights(weights map[string]float64) map[string]float64 {
	result := make(map[string]float64, len(weights))
	for k, v := range weights {
		result[k] = v
	}
	return result
}

func copyProbs(probs map[string]float64) map[string]float64 {
	result := make(map[string]float64, len(probs))
	for k, v := range probs {
		result[k] = v
	}
	return result
}
